src/lib/yang/util.lua

```diff
@@ -87,12 +87,47 @@ function ipv4_ntop(addr)
    return ipv4:ntop(ffi.new('uint32_t[1]', lib.htonl(addr)))
 end
 
-function string_output_file()
+ffi.cdef [[
+void* malloc (size_t);
+void free (void*);
+]]
+
+function string_io_file()
+   local function alloc(n)
+      return ffi.gc(ffi.cast('char*', ffi.C.malloc(n)), ffi.C.free)
+   end
+
    local file = {}
-   local out = {}
-   function file:write(str) table.insert(out, str) end
-   function file:flush(str) return table.concat(out) end
-   function file:clear(str) out = {} end
+   local size = 1024
+   local buf = alloc(size)
+   local written = 0
+   local read = 0
+   function file:write(str)
+      while size - written < #str do
+         if 0 < read then
```

Comment:
I have a doubt here. I think the `else` branch is always taken. The `then` branch is taken when read is greater than 0, but `read` is initially as zero and when it gets updated (in `write` and `reset`) is set to zero. So what actually happens in the end is that the else branch is always taken which doubles `size` until the whole string fits and then the string is copied to memory. Maybe I'm misreading the code, in what condition the `then` branch would be taken?

Reply:
You are right!  I think I was intending to add a method to consume bytes from the front.  That would still be a nice optimization in some cases, so let's maybe keep the code like it is.